chore: adjust comparison docs for claude agents sdk (#3612)
langchain-ai/docs: src/oss/deepagents/comparison.mdx

---
title: Comparison with Claude Agent SDK
sidebarTitle: Comparison with Claude Agent SDK
description: Compare LangChain Deep Agents with the Claude Agent SDK to choose the right tool for your use case.
---

This page explains how [LangChain Deep Agents](/oss/deepagents/overview) compares to the [Claude Agent SDK](https://platform.anthropic.com/docs/en/agent-sdk/overview). Both are harnesses for building custom agents, but they make different tradeoffs around execution environments, deployment, and vendor coupling.

<Info>
    Deep Agents is used in production by [OpenSWE](https://github.com/langchain-ai/open-swe) and [LangSmith Fleet](/langsmith/fleet/index).
</Info>

## At a glance

| | **Deep Agents** | **Claude Agent SDK** |
| --- | --- | --- |

> | **Where the agent runs** | Inside a sandbox, or outside a sandbox executing commands remotely | Inside a sandbox only |

Reviewer: ```suggestion
| **Where the agent runs** | Inside a sandbox, or outside a sandbox executing commands remotely | Inside a sandbox |
```

| **Execution backend** | Pluggable: [local, virtual filesystem, remote sandbox, or custom](/oss/deepagents/backends) | Local filesystem of the sandbox it runs in |

> | **Model provider** | Any (Anthropic, OpenAI, Google, 100+ others) | Claude only (Anthropic, Bedrock, Vertex, Azure) |

Reviewer: ```suggestion
| **Model provider** | Any (Anthropic, OpenAI, Google, 100+ others) | Claude (Anthropic, Bedrock, Vertex, Azure) |
```

> | **Deployment** | [LangSmith managed cloud](/oss/langgraph/deploy) via [`deepagents deploy`](/oss/deepagents/deploy), or self-host a [standalone image](/langsmith/deploy-standalone-server) via [`langgraph build`](/langsmith/cli#build) | [Self-host](https://code.claude.com/docs/en/agent-sdk/hosting) only. You build the server, auth, and streaming layer. [Claude managed agents](https://platform.claude.com/docs/en/managed-agents/overview) is a separate product |

Reviewer: ```suggestion
| **Deployment** | [LangSmith managed cloud](/oss/langgraph/deploy) via [`deepagents deploy`](/oss/deepagents/deploy), or self-host a [standalone image](/langsmith/deploy-standalone-server) via [`langgraph build`](/langsmith/cli#build) | [Self-host](https://code.claude.com/docs/en/agent-sdk/hosting). You build the server, auth, and streaming layer. [Claude managed agents](https://platform.claude.com/docs/en/managed-agents/overview) is a separate product |
```

| **Multi-tenancy** | [Built-in](/oss/deepagents/going-to-production#multi-tenancy): scoped threads, per-user sandboxes, RBAC | Build it yourself |
| **License** | MIT | MIT (Claude Code itself is proprietary) |

> ## What Deep Agents gives you over the Claude Agent SDK

Reviewer: We should try to keep this balanced. Are there any points you've found in favour of Claude Agent? If so let's rename this and cover both
```suggestion
## Main differences
```

> ### Decoupled agent and execution environment

Reviewer: ```suggestion
### Agent and execution environment
```

> There are [two patterns for connecting agents to sandboxes](https://www.langchain.com/blog/the-two-patterns-by-which-agents-connect-sandboxes): running the agent *inside* the sandbox, or running the agent outside and **using the sandbox as a tool**. The Claude Agent SDK only supports the first. Your agent runs inside a sandbox and executes tools against that sandbox's local filesystem.

Reviewer: ```suggestion
There are [two patterns for connecting agents to sandboxes](https://www.langchain.com/blog/the-two-patterns-by-which-agents-connect-sandboxes): running the agent *inside* the sandbox, or running the agent outside and **using the sandbox as a tool**.

The Claude Agent SDK supports the first. Your agent runs inside a sandbox and executes tools against that sandbox's local filesystem.
```

Reviewer 2: We can be more forceful on decoupled agent and execution part since Anthropic itself runs Claude managed agents in a decoupled model. What about something like this -- "The Claude Agent SDK only supports the first. Your agent runs inside a sandbox and executes tools against the sandbox's local filesystem. Notably, Claude managed agents (Anthropic's hosted model) use a decoupled model, which reflects where production agent architectures are heading. Deep Agents supports both patterns today, and lets you pick a backend to wire them together."

Reviewer: ```suggestion
There are [two patterns for connecting agents to sandboxes](https://www.langchain.com/blog/the-two-patterns-by-which-agents-connect-sandboxes): running the agent *inside* the sandbox, or running the agent outside and **using the sandbox as a tool**.

The Claude Agent SDK only supports the first. Your agent runs inside a sandbox and executes tools against the sandbox's local filesystem. Anthropic's hosted model [Claude managed agents](https://platform.claude.com/docs/en/managed-agents/overview) use a decoupled model, which reflects where production agent architectures are heading.
```

Deep Agents supports both, and lets you pick a [backend](/oss/deepagents/backends#quickstart) to wire them together. In practice, this means you can:

- Run the agent inside a sandbox (same model as Claude Agent SDK).
- Run the agent in a long-lived container and [use a remote sandbox as a tool](https://www.langchain.com/blog/the-two-patterns-by-which-agents-connect-sandboxes), executing commands over the network.
- Swap in a virtual filesystem for tests, or a custom backend for your own infrastructure.

> ### Multi-tenancy without the plumbing

Reviewer: ```suggestion
### Multi-tenancy
```

> This becomes a real problem when you productionize a Claude Agent SDK app and expose it to many end users. Because the SDK ties the agent to its sandbox, giving each user an isolated execution environment requires building an API wrapper that spins up a sandbox per user, tracks which sandbox belongs to whom, and tears it down afterwards.

Reviewer: ```suggestion
When you productionize your application, you generally expose it to many end users and must isolate the environtment for each user.

In Claude Agent SDK, the SDK ties the agent to its sandbox. To give each user an isolated execution environment, you must build an API wrapper that spins up a sandbox per user, tracks which sandbox belongs to whom, and tears it down afterwards.
```

Deep Agents handles this directly: configure a sandbox [per user or per assistant](/oss/deepagents/going-to-production#lifecycle) in the harness, with scoped threads, run history, and [RBAC](<link to deepagents auth docs>) included. If you use [LangSmith Sandbox](/langsmith/sandbox-auth-proxy#sandbox-auth-proxy), you also get an auth proxy out of the box so end users can call third-party APIs from the sandbox without you provisioning credentials per user.

> ### A production agent server out of the box

Reviewer: ```suggestion
### A production agent server
```

To expose a [self-hosted Claude Agent SDK](https://code.claude.com/docs/en/agent-sdk/hosting) app to end users, you write your own HTTP/WebSocket or SSE server that invokes the agent, streams tokens back, and manages conversation threads. That server is yours to build, operate, and secure.

Deep Agents deployments include an [agent server](/langsmith/agent-server) out of the box: streaming endpoints, thread management, run history, webhooks, and [authentication](<link to deepagents auth docs>).

> ### Managed cloud or self-hosted, your choice

Reviewer: ```suggestion
### Managed cloud or self-hosted
```

Claude Agent SDK deployments are [self-hosted](https://code.claude.com/docs/en/agent-sdk/hosting). The SDK and [Claude managed agents](https://platform.claude.com/docs/en/managed-agents/overview) are separate products. Code written against the SDK does not deploy directly to the managed offering.

Deep Agents runs in two modes without code changes:

- **Managed:** deploy to [LangSmith managed cloud](/oss/langgraph/deploy) with [`deepagents deploy`](/oss/deepagents/deploy).
- **Self-hosted:** run [`langgraph build`](/langsmith/cli#build) to produce a [standalone Docker image](/langsmith/deploy-standalone-server) you can deploy anywhere.

> ### No vendor lock-in

Reviewer: ```suggestion
### LLM
```

> Claude Agent SDK is tightly coupled to Claude models and Anthropic's execution model. With Deep Agents, you pick the model provider, the execution backend, and the deployment target independently, so your harness choice doesn't lock in your model or your infrastructure.

Reviewer: ```suggestion
Claude Agent SDK is tightly coupled to and optimized for Claude models and Anthropic's execution model. 

With Deep Agents, you pick the model provider, the execution backend, and the deployment target independently, so your harness choice doesn't lock in your model or your infrastructure.
```

Reviewer 2: What does Anthropic's execution model mean? Every harness is going to have an execution model. I think we should be more specific with something like: 'Claude Agent SDK execution bundles the model, backend, and deployment, with neither able to be independently swapped.' Something like this makes the DA decoupled approach clearer.

Reviewer: ```suggestion
Claude Agent SDK execution bundles the model, backend, and deployment and optimizes support between all three.

With Deep Agents, you pick the model provider, the execution backend, and the deployment target independently. By choosing this harness you retain maximum flexibility in your choice of model and infrastructure.
```

## Summary

- **Choose Deep Agents** if you want model and infrastructure flexibility, built-in multi-tenant deployment, and the option to run managed or self-hosted without code changes.

> - **Choose Claude Agent SDK** if you're committed to Claude, are happy to self-host, and are willing to build the API, auth, and multi-tenant layers yourself.

Reviewer: ```suggestion
- **Choose Claude Agent SDK** if you're committed to Claude, and wish to self-host, and build the API, auth, and multi-tenant layers yourself.
```

Reviewer 2: The "Choose Claude Agent SDK" is a bit thin and it makes the comparison read more biased towards our own product. Is there something more concrete that Claude Agent SDK actually does better? For instance, is it faster to get started? better documentation? smaller operational surface?

<Note>
    **Notice a mistake?**

>     We drafted this comparison on April 16th, 2026. If products have changed or you spot an error, please [file an issue](https://github.com/langchain-ai/docs/issues).

Reviewer 2: It's my first time seeing one of these explicit "if you notice a mistake" sections. Is this typical? It reads as if we're unsure of the information we're providing, and we already let readers file an issue below that if they notice a mistake.

Reviewer: I think it makes it more trustworthy to point out we're willing to make changes but I'll amend to remove the mistake part
```suggestion
    We drafted this comparison on April 16th, 2026. If products have changed, please [file an issue](https://github.com/langchain-ai/docs/issues).
```
